internal/kafka/internal/config/kafka_supported_instance_types.go

```diff
@@ -0,0 +1,202 @@
+	QuotaConsumed               int      `yaml:"quotaConsumed"`
+	QuotaType                   string   `yaml:"quotaType"`
+	CapacityConsumed            int      `yaml:"capacityConsumed"`
+}
+
+// validates Kafka instance size configuration to ensure the following:
+//
+// - all properties must be defined
+// - any non-id string values must be parseable
+// - any int values must not be less than or equal to zero
+func (k *KafkaInstanceSize) validate(instanceTypeId string) error {
+	if k.EgressThroughputPerSec.IsEmpty() || k.IngressThroughputPerSec.IsEmpty() ||
+		k.MaxDataRetentionPeriod == "" || k.MaxDataRetentionSize.IsEmpty() || k.Id == "" || k.QuotaType == "" {
+		return fmt.Errorf("Kafka instance size '%s' for instance type '%s' is missing required parameters.", k.Id, instanceTypeId)
+	}
+
+	egressThroughputQuantity, err := k.EgressThroughputPerSec.ToK8Quantity()
+	if err != nil {
+		return fmt.Errorf("egressThroughputPerSec for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	ingressThroughputQuantity, err := k.IngressThroughputPerSec.ToK8Quantity()
+	if err != nil {
+		return fmt.Errorf("ingressThroughputPerSec for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	maxDataRetentionSize, err := k.MaxDataRetentionSize.ToK8Quantity()
+	if err != nil {
+		return fmt.Errorf("maxDataRetentionSize for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	maxDataRetentionPeriod, err := duration.ParseISO8601(k.MaxDataRetentionPeriod)
+	if err != nil {
+		return fmt.Errorf("maxDataRetentionPeriod for Kafka instance type '%s', size '%s' is invalid: %s", k.Id, instanceTypeId, err.Error())
+	}
+
+	if maxDataRetentionPeriod.IsZero() || egressThroughputQuantity.CmpInt64(1) < 0 ||
+		ingressThroughputQuantity.CmpInt64(1) < 0 || maxDataRetentionSize.CmpInt64(1) < 0 ||
+		k.TotalMaxConnections <= 0 || k.MaxPartitions <= 0 || k.MaxConnectionAttemptsPerSec <= 0 ||
+		k.QuotaConsumed < 1 || k.CapacityConsumed < 1 {
+		return fmt.Errorf("Kafka instance size '%s' for instance type '%s' specifies a property value less than or equals to Zero.", k.Id, instanceTypeId)
+	}
+
+	return nil
+}
+
+type SupportedKafkaInstanceTypesConfig struct {
+	SupportedKafkaInstanceTypes []KafkaInstanceType `yaml:"supported_instance_types"`
+}
+
+func (s *SupportedKafkaInstanceTypesConfig) validate() error {
+	existingInstanceTypes := make(map[string]int, len(s.SupportedKafkaInstanceTypes))
+
+	for _, KafkaInstanceType := range s.SupportedKafkaInstanceTypes {
+		if _, ok := existingInstanceTypes[KafkaInstanceType.Id]; ok {
+			return fmt.Errorf("Kafka instance type id '%s' was defined more than once.", KafkaInstanceType.Id)
+		}
+		existingInstanceTypes[KafkaInstanceType.Id]++
+
+		if err := KafkaInstanceType.validate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type KafkaSupportedInstanceTypesConfig struct {
```

Comment:
same as above, can struct definitions be placed at the top

Reply:
Same comment as https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/965#discussion_r859546733